### PR TITLE
Travis setup

### PR DESCRIPTION
**Pull request by jb803**
Added everything required for Travis CI to automatically build and test the code. I've assumed we want to stick with installing FEniCS via anaconda. 

Each Travis job:

1. Downloads miniconda
2. Installs FEniCS, pytest, sphinx and black under the specified python version (just 3.6 at the moment)
3. Runs the relevant test suite.

Tox doesn't play well with anaconda so there's now a test script, `provision/travis/run-test.sh` , which runs different tests based on the `TEST_ENV` environment variable. If `TEST_ENV` is not passed then it will run all the tests. 

The current tests are:
1. `pytest` on the `test/` folder. I've filtered out `DepreciationWarning` and `PendingDepreceationWarning` because it turns out `import dolfin` causes about a billion of these to be produced (who would have thought it? :p)
2. Tests building the doucmentation via sphinx
3. `black --check tonic/`  . Checks to make sure `black` is happy with all the module code in `tonic/`. I don't know if we want to add `test/` to the files checked by `black`.  What do people think?

You don't seem to be able to get activate Travis on a repository until the `.travis.yml` file is present in the master branch. When this is merged, I'll enable Travis and then make the Travis checks a requirement for merging pull requests.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,7 @@
 /doc/build/*
 #ignore pip install info
 tonic.egg-info/*
+#ignore coverage info
+.coverage
+#ignore pytest chace
+.pytest_cache/*
```
Comment (Saintis) [inline .gitignore]: Needs line break

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  - conda info -a
+  - conda create -q -n test-environment -c conda-forge python=$TRAVIS_PYTHON_VERSION fenics pytest pytest-cov coveralls black sphinx
+  - source activate test-environment
+  #We install the package
+  - pip install -e .
+
+script: provision/travis/run-tests.sh
+after_success:
+  - coveralls
+
```
Comment (Saintis) [inline .travis.yml]: way too many line breaks

File: provision/travis/run-tests.sh [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+    black --check tonic/
+elif [ "$TEST_ENV" = "py36-test" ]; then
+    pytest --cov=tonic/ tonic test
+elif [ "$TEST_ENV" = "doc" ]; then
+    sphinx-build -b html doc/source/ doc/build/
+elif [ "$TEST_ENV" = "black" ]; then
+     black --check tonic/
+else
+    echo Uknown test environment
+fi
```
Comment (Saintis) [inline provision/travis/run-tests.sh]: needs line break

File: provision/travis/run-tests.sh [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+#No test environment is set, we run all the tests, build the docs and do black
+    pytest --cov=tonic/ tonic test
+    sphinx-build -b html doc/source/ doc/build/
+    black --check tonic/
+elif [ "$TEST_ENV" = "py36-test" ]; then
+    pytest --cov=tonic/ tonic test
+elif [ "$TEST_ENV" = "doc" ]; then
+    sphinx-build -b html doc/source/ doc/build/
+elif [ "$TEST_ENV" = "black" ]; then
+     black --check tonic/
```
Comment (Saintis) [inline provision/travis/run-tests.sh]: some repeated lines of code, is it possible to rewrite this so the sphinx, pytest and black lines are only written once?

File: test/test_deps.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+"""Tests  various dependencies. This is an initial example test required to allow pytest to be happy when there are no actual tests written """
+
+def test_importDolfin():
+    import dolfin as dolf
```
Comment (Saintis) [inline test/test_deps.py]: needs line break

File: provision/travis/run-tests.sh (hunk 4 above)
Comment (Saintis) [reply provision/travis/run-tests.sh]: also do black on whole project, including tests.

**Review by Saintis — CHANGES_REQUESTED** [review]

File: provision/travis/run-tests.sh (hunk 4 above)
Comment (jb803) [reply provision/travis/run-tests.sh]: Test invocations are declared at beginning of the bash file now.
black runs also checks the test/ directory

**Review by Saintis — APPROVED** [review]

**Pull request merged**